### PR TITLE
Implement a simple ban manager

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
Currently it just replaces the existing usage of banning addresses (by banning the local IP).

This will be expanded to use the onFailedRequest() API later.

I want to separate all of this stuff into digestible PRs.

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
Did a couple of pushes. Renamed to `BanManager`.

**Comment by AndrejMitrovic** [comment]
Improved the API a bit (and added tests).

Now there is `banUntil` and `banFor`, which ban until a specific time, or ban an address for the specified time starting from the current time.

**Comment by AndrejMitrovic** [comment]
Updated with a small fix. If the node is banned, there's no need to keep track of failed requests.

File: source/agora/common/BanManager.d [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+module agora.common.BanManager;
+
+import agora.common.Data;
+
+import core.stdc.stdlib;
+import core.stdc.time;
+
+/// ditto
+class BanManager
```
Comment (Geod24) [inline source/agora/common/BanManager.d]: `public`

File: source/agora/common/BanManager.d [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    ***************************************************************************/
+
+    protected time_t getCurTime () const
+    {
+        return time(null);
+    }
+
+    /***************************************************************************
+
+        Workaround for `ref` returns which store a copy to
```
Comment (Geod24) [inline source/agora/common/BanManager.d:164]: The title should say what the function does. The description can explain why.

**Comment by AndrejMitrovic** [comment]
Updated with suggested changes.

**Review by Geod24 — APPROVED** [review]

**Pull request merged**